ticker/src/main/java/com/robinhood/ticker/TickerColumnManager.java

```diff
@@ -77,25 +77,47 @@ boolean shouldDebounceText(char[] text) {
     /**
      * Tell the column manager the new target text that it should display.
      */
-    boolean setText(char[] text, boolean animate) {
+    void setText(char[] text, boolean animate) {
         if (characterList == null) {
             throw new IllegalStateException("Need to call setCharacterList(char[]) first.");
         }
 
-        final int newTextSize = text.length;
-        if (animate) {
-            // If we are animating, we don't want to remove old columns yet.
-            insertColumnsUpTo(newTextSize);
-        } else {
-            ensureColumnSize(newTextSize);
+        // First remove any zero-width columns
+        for (int i = 0; i < tickerColumns.size(); ) {
+            final TickerColumn tickerColumn = tickerColumns.get(i);
+            if (tickerColumn.getCurrentWidth() > 0) {
+                i++;
+            } else {
+                tickerColumns.remove(i);
+            }
         }
 
-        final int columnsSize = tickerColumns.size();
-        for (int i = 0; i < columnsSize; i++) {
-            tickerColumns.get(i).setTargetChar(i < newTextSize ? text[i] : TickerUtils.EMPTY_CHAR);
+        // Use Levenshtein distance algorithm to figure out how to manipulate the columns
+        final int[] actions = LevenshteinUtils.computeColumnActions(getCurrentText(), text);
+        int columnIndex = 0;
+        int textIndex = 0;
+        for (int i = 0; i < actions.length; i++) {
+            switch (actions[i]) {
+                case LevenshteinUtils.ACTION_INSERT:
+                    tickerColumns.add(columnIndex,
+                            new TickerColumn(characterList, characterIndicesMap, metrics));
+                case LevenshteinUtils.ACTION_SAME:
+                    tickerColumns.get(columnIndex).setTargetChar(text[textIndex]);
+                    columnIndex++;
+                    textIndex++;
+                    break;
+                case LevenshteinUtils.ACTION_DELETE:
+                    if (animate) {
```

Comment:
Yeah you are right, the `TickerColumnManager` actually do not need to know about `animate` flag anymore as it can (and should) simply set the target state of each column (e.g. animate to EMPTY_CHAR to shrink to 0 width), and then the caller can set animation state via `setAnimationProgress`.